---
title: "R Assignment"
output: html_notebook
---
Setting the directory
```{r}
setwd("~/Desktop/BioDataSkillsClass/UNIX_Assignment")
```

Loading files to R

```{r}
fang <- read.table("fang_et_al_genotypes.txt", header = TRUE )
position <- read.table("snp_position.txt", header = TRUE, fill = TRUE)
```
#Part 1

load packages
```{r}
library(dplyr)
library(gtools)
```

##Data Inspection
```{r}
str(fang)
str(position)
dim(fang)
dim(position)
nrow(fang)
nrow(position)
ncol(fang)
ncol(position)

head(fang, n = 1)
head(position, n = 3)
```

##Data processing

Subset necessary columns
```{r}
fang_subset <- subset(fang, select = -c(1:2))
```

Extracting maize and teosinte data
```{r}
maize <- c("ZMMIL","ZMMLR","ZMMMR")
maize_extracted <- fang_subset[fang_subset$Group %in% maize,]

teosinte <- c("ZMPBA", "ZMPIL", "ZMPJA")
teosinte_extracted <- fang_subset[fang_subset$Group %in% teosinte,]
```

Transpose the data
```{r}
maize_transposed <- as.data.frame(t(maize_extracted[,-1]))

teosinte_transposed <- as.data.frame(t(teosinte_extracted[,-1]))
```

Subset columns of snp_position file that matter
```{r}
position_subset <- subset(position, select = c(1,3,4))
```

Sort files
```{r}
maize_sorted <- maize_transposed[order(row.names(maize_transposed)),]
teosinte_sorted <- teosinte_transposed[order(row.names(teosinte_transposed)),]
position_sorted <- position_subset[order(position_subset$SNP_ID),]
```

Remove data from "unknown and random" chromosomes in SNP position file
```{r}
chrom <- c(1,2,3,4,5,6,7,8,9,10)
position_sorted_selected <- subset(position_sorted, Chromosome %in% chrom)
position_sorted <- position_sorted_selected[-1,]
```
Making row names as columns so I can join the data frames
```{r}
maize_row <- add_rownames(maize_sorted, "SNP_ID")
teosinte_row <- add_rownames(teosinte_sorted, "SNP_ID")
```

Merge maize and teosinte data with snp position data
```{r}
maize_merged <- merge(position_sorted, maize_row, by="SNP_ID")
teosinte_merged <- merge(position_sorted, teosinte_row, by="SNP_ID")
``` 

Sort by SNP position - ascending order
```{r}
maize_sorted_by_position <- arrange(maize_merged, Position)  #ascending position
teosinte_sorted_by_position <- arrange(teosinte_merged, Position) #ascending position 

maize_sorted <- order_by()
```

Substitute symbols - no data by ?
```{r}
gsub("","?", maize_sorted_by_position) 
gsub("", "?", teosinte_sorted_by_position)
```

Sort by SNP position - descending order
```{r}
maize_desc_sorted_by_position <- arrange(maize_merged, desc(Position))
teosinte_desc_sorted_by_position <- arrange(teosinte_merged, desc(Position))
```
Substitute symbols - no data by -
 
```{r}
gsub("","-", maize_desc_sorted_by_position) 
gsub("","-", teosinte_desc_sorted_by_position)
```

Extract data by chromosome number
```{r}
for(i in 1:10){
 temp_df<-maize_sorted_by_position[maize_sorted_by_position[,2]==i,]
 write.csv(temp_df, paste("maize_chr",i,".csv",sep=""),row.names=F)}

for(i in 1:10){
 temp_df<-teosinte_sorted_by_position[teosinte_sorted_by_position[,2]==i,]
 write.csv(temp_df, paste("teosinte_chr",i,".csv",sep=""),row.names=F)}

for(i in 1:10){
 temp_df<-maize_desc_sorted_by_position[maize_desc_sorted_by_position[,2]==i,]
 write.csv(temp_df, paste("maize_desc_chr",i,".csv",sep=""),row.names=F)}

for(i in 1:10){
 temp_df<-teosinte_desc_sorted_by_position[teosinte_desc_sorted_by_position[,2]==i,]
 write.csv(temp_df, paste("teosinte_desc_chr",i,".csv",sep=""),row.names=F)}
```


###Part 1 finished

#Part 2

Tidying the data 
```{r}
library(reshape2)

maize_reshaped <- melt(maize_merged, id=(c("SNP_ID","Chromosome","Position")))

teosinte_reshaped <- melt(teosinte_merged, id=(c("SNP_ID","Chromosome","Position")))

totalSNPs_maize <- with(maize_reshaped, aggregate(cbind(SNP_ID) ~ Chromosome, FUN=function(x){length(unique(x))}))

totalSNPs_teosinte <- with(teosinte_reshaped, aggregate(cbind(SNP_ID) ~ Chromosome, FUN=function(x){length(unique(x))}))
```

Plot total number of SNPs by chromosome number
```{r}
library(ggplot2)

ggplot(data = totalSNPs_maize, aes(x=Chromosome, y=SNP_ID)) + geom_bar(stat="identity")

ggplot(data = totalSNPs_teosinte, aes(x=Chromosome, y=SNP_ID)) + geom_bar(stat="identity")
```

Create new column indicating if site is homozygous or heterozygous
```{r}
maize_reshaped$heterozygosity <- if (maize_reshaped$value == "A/A") {
  print(homozygous)
} else if (maize_reshaped$value == "C/C") {
  print(homozygous)
} else if (maize_reshaped$value == "G/G") {
  print(homozygous)
} else if (maize_reshaped$value == "T/T") {
  print(homozygous)
} else {
  print(heterozygous)
}

teosinte_reshaped$heterozygosity <- if (teosinte_reshaped$value == "A/A") {
  print(homozygous)
} else if (teosinte_reshaped$value == "C/C") {
  print(homozygous)
} else if (teosinte_reshaped$value == "G/G") {
  print(homozygous)
} else if (teosinte_reshaped$value == "T/T") {
  print(homozygous)
} else {
  print(heterozygous)
}

##Recode this section
```

Substitute missing data by NA
```{r}
gsub("?/?","NA", maize_reshaped) 
gsub("?/?","NA", teosinte_reshaped)
```

Sort dataframe by Species_ID and Group values
```{r}
final_maize <- arrange(maize_reshaped, Species_ID, Group)

final_teosinte <- arrange(teosinte_reshaped, Species_ID, Group)
```

Graph that shows the proportion of homozygous and heterozygous sites as well as missing data in each species
```{r}
ggplot(data = final_maize, aes(x=Group, y=heterozygosity)) + geom_bar(position="stack")

ggplot(data = final_teosinte, aes(x=Group, y=heterozygosity)) + geom_bar(position="stack")
```
